From 5333b2e8febb3e54eba0b2aecd7a182442b6b933 Mon Sep 17 00:00:00 2001
From: Aaron Williams <aaron.williams@cavium.com>
Date: Mon, 10 Sep 2018 16:54:03 -0700
Subject: [PATCH 0795/1239] octeontx2: mmc: Disable tap calibration in ASIM

This patch adds code to detect when U-Boot is running on ASIM and to
disable the tap calibration procedure which ASIM does not support.

Signed-off-by: Aaron Williams <aaron.williams@cavium.com>
---
 .../include/asm/arch-octeontx/octeontx_mmc.h  |  3 +++
 drivers/mmc/octeontx_mmc.c                    | 27 +++++++++++++++++++
 2 files changed, 30 insertions(+)

diff --git a/arch/arm/include/asm/arch-octeontx/octeontx_mmc.h b/arch/arm/include/asm/arch-octeontx/octeontx_mmc.h
index f84f94a583..b0d62fe523 100644
--- a/arch/arm/include/asm/arch-octeontx/octeontx_mmc.h
+++ b/arch/arm/include/asm/arch-octeontx/octeontx_mmc.h
@@ -165,6 +165,9 @@ struct octeontx_mmc_slot {
 	bool		power_active_high:1;
 	bool		ro_inverted:1;	/** True if write-protect is inverted */
 	bool		cd_inverted:1;	/** True if card-detect is inverted */
+#if defined(CONFIG_ARCH_OCTEONTX2)
+	bool		is_asim:1;	/** True if we're running in ASIM */
+#endif
 };
 
 struct octeontx_mmc_host {
diff --git a/drivers/mmc/octeontx_mmc.c b/drivers/mmc/octeontx_mmc.c
index 57db77c4e0..7ea797da7f 100644
--- a/drivers/mmc/octeontx_mmc.c
+++ b/drivers/mmc/octeontx_mmc.c
@@ -4450,6 +4450,10 @@ static int octeontx2_mmc_calibrate_delay(struct mmc *mmc)
 	union cavm_mio_emm_tap emm_tap;
 	ulong start = get_timer(0);
 
+	/* ASIM doesn't support this */
+	if (slot->is_asim)
+		return 0;
+
 	emm_calb.u = 0;
 	emm_calb.s.start = 1;
 	mmc_write_csr(mmc, CAVM_MIO_EMM_CALB(), emm_calb.u);
@@ -4505,6 +4509,10 @@ static int octeontx_mmc_set_bus_timing(struct mmc *mmc)
 	union cavm_mio_emm_timing timing;
 	int delay;
 
+	/* ASIM doesn't support this. */
+	if (slot->is_asim)
+		return 0;
+
 	timing.u = 0;
 	delay = octeontx2_mmc_calc_delay(mmc, slot->cmd_clk_skew);
 	if (delay < 0) {
@@ -5469,6 +5477,10 @@ int octeontx_mmc_initialize(struct udevice *dev)
 	int repeat;
 	int found = 0;
 	int slot_index;
+#if defined(CONFIG_ARCH_OCTEONTX2)
+	bool is_asim = false;
+	const char *board_model;
+#endif
 
 	debug("%s ENTER\n", __func__);
 
@@ -5504,9 +5516,24 @@ int octeontx_mmc_initialize(struct udevice *dev)
 		return -1;
 	}
 
+#if defined(CONFIG_ARCH_OCTEONTX2)
+	/* If we're running in ASIM then we need to skip the tap calibration
+	 * and just use the default settings.  Look in the fdt at
+	 * /cavium,bdk/BOARD-MODEL.  It will start with "ASIM-"
+	 */
+	board_model = fdt_getprop(gd->fdt_blob,
+				  fdt_path_offset(gd->fdt_blob, "/cavium,bdk"),
+				  "BOARD-MODEL", NULL);
+	if (board_model && !strncmp(board_model, "ASIM-", 5))
+		is_asim = true;
+
+#endif
 	for (slot_index = 0; slot_index < OCTEONTX_MAX_MMC_SLOT;
 	     slot_index++) {
 		slot = &host->slots[slot_index];
+#if defined(CONFIG_ARCH_OCTEONTX2)
+		slot->is_asim = is_asim;
+#endif
 		mmc = slot->mmc;
 		if (!mmc)
 			continue;
-- 
2.29.0

